src/Intervals.php

```diff
@@ -0,0 +1,333 @@
+<?php
+
+/*
+ * This file is part of composer/semver.
+ *
+ * (c) Composer <https://github.com/composer>
+ *
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+namespace Composer\Semver;
+
+use Composer\Semver\Constraint\Constraint;
+use Composer\Semver\Constraint\ConstraintInterface;
+use Composer\Semver\Constraint\EmptyConstraint;
+use Composer\Semver\Constraint\MultiConstraint;
+
+/**
+ * Helper class to evaluate constraint by compiling and reusing the code to evaluate
+ */
+class Intervals
+{
+    private static $intervalsCache = array();
+
+    /**
+     * @psalm-var array<string, int>
+     */
+    private static $opSortOrder = array(
+        '>=' => -3,
+        '<' => -2,
+        '==' => -1,
+        '!=' => 1,
+        '>' => 2,
+        '<=' => 3,
+    );
+
+    public static function clear()
+    {
+        self::$intervalsCache = array();
+    }
+
+    public static function isSubsetOf(ConstraintInterface $candidate, ConstraintInterface $constraint)
+    {
+        if ($constraint instanceof EmptyConstraint) {
+            return true;
+        }
+
+        $intersectionIntervals = self::get(new MultiConstraint(array($candidate, $constraint), true));
+        $candidateIntervals = self::get($candidate);
+        if (\count($intersectionIntervals['intervals']) !== \count($candidateIntervals['intervals'])) {
+            return false;
+        }
+
+        foreach ($intersectionIntervals['intervals'] as $index => $interval) {
+            if (!isset($candidateIntervals['intervals'][$index])) {
+                return false;
+            }
+
+            if ((string) $candidateIntervals['intervals'][$index]['start'] !== (string) $interval['start']) {
+                return false;
+            }
+
+            if ((string) $candidateIntervals['intervals'][$index]['end'] !== (string) $interval['end']) {
+                return false;
+            }
+        }
+
+        if (\count($intersectionIntervals['devConstraints']) !== \count($candidateIntervals['devConstraints'])) {
+            return false;
+        }
+        foreach ($intersectionIntervals['devConstraints'] as $index => $c) {
+            if ((string) $c !== (string) $candidateIntervals['devConstraints'][$index]) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    public static function isIntersectionOf(ConstraintInterface $a, ConstraintInterface $b)
```

Comment:
the name looks weird to me. This checks whether the 2 constraints are intersecting (they have a non empty intersection), not whether one of them is an intersection of the other (that does not make any sense).

Reply:
indeed, `intersectsWith` is probably better

Reply:
Yup true, I had it as `intersects` at first but then thought it sounded weird next to isSubsetOf. Then we also have `get` which I guess should be `getIntervals` for clarity as this grew to have more than just one getter.

I'd say `haveIntersection` if we return bool probably makes sense here. or `isIntersecting` as in `$a is intersecting $b`? The plural form kinda sounds odd.

Reply:
The class is plural `Intervals` so I think the most logical sounding is `lntervals::haveIntersections` ?